darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py

```diff
@@ -0,0 +1,442 @@
+"""
+Module of data pre-processing functions for constructing the heatmap figure.
+"""
+from typing import Dict, Any, Tuple, Sequence, TypedDict
+
+import pandas as pd
+import numpy as np
+
+
+class SegDict(TypedDict):
+    """
+    Custom type hint class for `dict_list` argument in `get_rd_wr_dfs()`.
+    """
+
+    id: int
+    rank: int
+    hostname: str
+    write_count: int
+    read_count: int
+    write_segments: pd.DataFrame
+    read_segments: pd.DataFrame
```

Comment:
Can you clarify what this `SegDict` class does for static type checking? For example, if I modify it arbitrarily:

```diff
diff --git a/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py b/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
index c2bd5c1..4bdd93f 100644
--- a/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
+++ b/darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py
@@ -19,6 +19,7 @@ class SegDict(TypedDict):
     read_count: int
     write_segments: pd.DataFrame
     read_segments: pd.DataFrame
+    stuff: int
```

`mypy` is still perfectly happy with the check on the `darshan` namespace as far as I can tell (apart from the usual suspects)

Reply:
Note that static type checking is probably pretty low on our priority list, but just chipping away at any complexity in the PR that I'm curious about.

Reply:
The purpose of the `TypedDict` is to specify the types of each individual key/value pair in a dictionary. Since the record dictionaries are using `int`/`str`/pandas dataframes I figured it was worth adding detail. Otherwise I believe the only option is to use `dict`, which is really `dict[Any, Any]`, so it doesn't check anything inside the record dictionary. 

But this is interesting. I got the same thing when I tried it. I expected it would detect any missing/extra keys as-is, but now that I think about it, I'm not sure how it would know what keys were available since we never actually specify them. 

For example, if we had something like:
```python 
from typing import TypedDict 

class SegDict(TypedDict):
    foo: int 
    bar: str 

def func(test_dict: SegDict):
    test_dict['foo'] += 10
```
How would it ever know that `bar` is missing? It would have to _assume_ it is missing because we never explicitly write down what keys `test_dict` has. One could argue that if it isn't used in the function, then it should be assumed missing. But maybe `mypy` doesn't make that assumption because that would require you to write a custom `TypedDict` object for each function, even if the same object is getting passed around. 

Now, if you alter the `read_segment` or `write_segment` entries (i.e. remove one, change its type, etc.) you do get errors. So it is adding value in that way at least. You can also remove any of the entries that aren't used without error, and you can change `rank` to any type you want without error. Again, I think this is because none of them are used in a way that can be tested so `mypy` has no way of knowing what they are (and doesn't assume they are missing).

Reply:
Adding to my last point -- if you change `read_segments` to `int`, the only reason it knows this is incorrect is because `int` doesn't have the `.size` or `.drop` methods. So I think they have to be used in a way that illustrates their type before `mypy` can make those determinations.

Reply:
The explanation in the [docs](https://mypy.readthedocs.io/en/stable/more_types.html#typeddict):

> A TypedDict object with extra items is compatible with (a subtype of) a narrower TypedDict, assuming item types are compatible...

Reply:
Anyway, this seems useful despite the limitations, so I'll mark this resolved, thanks!